### PR TITLE
[vendor] Add bazel vendor file for XKCP (i.e. standard KMAC impl)

### PR DESCRIPTION
**Pull request by ballifatih**
Follow-up of #16873 on vendoring XKCP.  This is an item on the list #16410. Again, with do-not-merge-yet warning.

I mostly followed the [example](https://github.com/lowRISC/opentitan/tree/master/third_party/freertos) @timothytrippel mentioned. Similar questions again:

- Are bazel scripts in the right place? Do they look sane?
- Is the example code generator in the right place `sw/device/lib/testing/kmac`?

The example code `kmac_generator_code.c` is not really necessary for this PR and I didn't yet implement the vector generator code. I am currently thinking about generating some C header files like [this](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/drivers/kmac_test_vectors.c), but feel free chime in. Thanks.

### CODE REVIEW

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:35]: Please 
* use `CHECK()` instead of `assert()` (also, `exit()` doesn't make a lot of sense in our environment since we are bare bones and don't have a host environment yet), and
* the `%!x` specifier instead of `print_long_array()`.

```suggestion
```

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:63]: nit: `CHECK_ARRAYS_EQ()`.

**Review by alphan — APPROVED** [review]
This mostly LGTM, just some nits but please wait for approval from @drewmacrae or @dmcardle or @milesdai for bazel related changes. Thanks @ballifatih !

File: third_party/xkcp/BUILD.xkcp.bazel
Comment (timothytrippel) [inline third_party/xkcp/BUILD.xkcp.bazel:25]: I think you may want to move some of the `.h`/`.inc` files listed here to the `srcs` list. See the [official Bazel documentation](https://bazel.build/reference/be/c-cpp#hdrs) on what to list where. To summarize, you want to place `#include`d files in the `srcs` list that are only included by header/source files that are specific to this library. If the header file is designed to be included by other targets that include this library as a dep (i.e., headers that make up the public API for this library), then you should put it in the `hdrs` list, otherwise it should go in `srcs`. I haven't looked into the source code of the lib you are vendoring in, but I'm guessing the `.inc` are examples of things that should go in `srcs`.

Comment (timothytrippel) [reply third_party/xkcp/BUILD.xkcp.bazel:25]: same applies to this lib below

**Review by timothytrippel — APPROVED** [review]
I have looked over the Bazel components, they mostly LGTM modulo one comment. Thanks @ballifatih !

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (timothytrippel) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:1]: Question: does this test vector generator need to be its own stand alone device program? How is it intended to be run?

Currently it is in `sw/device/` which would imply it is is designed to run on an opentitan device. However, it is a `cc_binary` target, which indicates it is a host binary.

If this is a standalone host tool then perhaps it should go under `sw/host/`. If this is suppose to be an on-device utility, then perhaps the `main` and helper functions should become testutils under `sw/device/lib/testing/kmac_testutils` where they can be used by tests that are run via the OTTF. Does that make sense?

**Comment by timothytrippel** [comment]
Dropped another comment regarding code organization. ^^

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (dmcardle) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:9]: Per the OpenTitan style guide, includes should generally be rooted at the repository root. https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#included-files

I admit this is a special circumstance because it's including a header from outside the OpenTitan repo, but I'd still try to use the full(er) path, `lib/high/Keccak/FIPS202/KeccakHash.h` and cut down on the `includes` in `third_party/xkcp/BUILD.xkcp.bazel`.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (dmcardle) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:50]: I think these could be strict aliasing violations. A `char*` is not necessarily compatible with `BitSequence*`. (Looks like `BitSequence` is a typedef of `uint8_t`, which is not guaranteed to be a character type.)

I think you might as well declare these as BitSequence arrays in the first place. For example:

```c
const BitSequence key[32] = {0x40, 0x41, 0x42, /* ... */ };
```

The other advantage of this approach is that it retains the array's size. Now, when you pass a pointer and length, you can use `ARRAYSIZE` rather than hardcoding the length.

File: third_party/xkcp/repos.bzl
Comment (dmcardle) [inline third_party/xkcp/repos.bzl:17]: Nit: typo in "restricted"

File: third_party/xkcp/update_main_license.patch
Comment (dmcardle) [inline third_party/xkcp/update_main_license.patch:9]: I don't think we should be stripping the license. Was there a reason for this?

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (dmcardle) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:60]: Paired with the above comment about declaring `const BitSequence key[32] = ...`, you can make the `keyBitLen` parameter `ARRAYSIZE(key) * 8` rather than hardcoding the length. Same goes for the other length parameters.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (dmcardle) [inline sw/device/lib/testing/kmac/kmac_vector_generate.c:57]: I would either delete this comment or replace it with a forward declaration.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (dmcardle) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:63]: @alphan This appears to be a host binary, so probably `CHECK_ARRAYS_EQ` doesn't make sense in this context?

File: sw/device/lib/testing/kmac/BUILD
Comment (dmcardle) [inline sw/device/lib/testing/kmac/BUILD:11]: I would rearchitect this into a library and add a `cc_test` target. That way, CI will warn you about compilation failures and unexpected results, if you encode expectations into the test!

If you intend to run it to extract results, you can either run the test directly or create a `cc_binary` that depends on the library and provides a `main()` function.

**Review by dmcardle — COMMENTED** [review]
Thanks for working on this @ballifatih! Leaving a few comments/questions to resolve.

File: third_party/xkcp/update_main_license.patch
Comment (dmcardle) [reply third_party/xkcp/update_main_license.patch:9]: On second thought, I'm guessing we're only using stuff contained in `libXKCP, UnitTests, Benchmarks nor KeccakSum`?

I'd probably call out the reasoning in a comment in the BUILD file that applies this patch.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (timothytrippel) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:9]: An example of how you would reference a header in a Bazel vendored external library is shown [here](https://github.com/lowRISC/opentitan/blob/aaad9d3fe4d8d66b2d84525d44f06f2b75543ec9/sw/device/lib/testing/test_framework/freertos_hooks.c#L6). I believe this would be `#include external/xkcp/lib/high/Keccak/FIPS202/KeccakHash.h`, implementing the pattern: `#include external/<bazel repo name>/path/to/header`.

To clarify, I don't think using the full include path will necessarily cut down ont he `includes` in `third_party/xkcp/BUILD.xkcp.bazel`, as those are needed to compile the `cc_libraries` in an external repository that may not follow the same code style guide, i.e., src files in the vendored in code might not use full include paths. For example, [it does not look like the the XKCP codebase uses full include paths.](https://github.com/XKCP/XKCP/blob/64404beeeb261b08a1076fe2f076e4e28dd9b040/lib/high/Keccak/SP800-185/SP800-185.c#L18)

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:35]: I think my mistake was to put this under `sw/device`, but this is actually host binary. Are these also available for host binaries?

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:63]: Yes, this is actually supposed to be host binary, my mistake was putting it under `sw/device`.

File: third_party/xkcp/BUILD.xkcp.bazel
Comment (ballifatih) [reply third_party/xkcp/BUILD.xkcp.bazel:25]: Thanks, makes sense. Fixing this.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:9]: I also thought about patching the header files of XKCP (so that they are using full paths), but then I thought maybe it a bit too much. I will make sure to use the full path for OT side code though. Thanks.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:50]: I will be using `uint8_t` arrays here. For the `ARRAYSIZE` macro, the definition comes from [a header in `/sw/device`](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/base/macros.h#L37). Just out of curiosity, since this will be a host code, can I use this header on the host side?

File: third_party/xkcp/update_main_license.patch
Comment (ballifatih) [reply third_party/xkcp/update_main_license.patch:9]: Commenting sounds reasonable. 

I think the bigger issue here is that I do not want and do not need to bring GPL-licensed parts of XKCP. I just could not figure out how to bring some parts XKCP library (without patching). This was easier with vendoring tool by exclusions.

The patching actually ends up bringing all the files to OT repo (see the content of `remove_restricted_files.patch`). Should I be using another tool to create and apply patches?

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:57]: Deleting sounds good, thanks.

File: sw/device/lib/testing/kmac/BUILD
Comment (ballifatih) [reply sw/device/lib/testing/kmac/BUILD:11]: I think what you said makes sense. I am not sure how to do this yet. Is converting `cc_binary` to `cc_test`, and adding some assertions to `kmac_vector_generate.c` does the trick? 

Or did you mean there should be some intermediate library `xyz`, which uses `@xkcp` as dependency as `xyz` is a dependency of `kmac_vector_generate`?

Eventually, it would be great if the test are automatically generated through XKCP and made available to KMAC library so that we can test this somewhere on CI. At least that's what I am hoping for.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c:1]: Having `kmac_vector_generate` as a host program makes more sense (therefore I will be moving it under `sw/host`). I just added this as an example use of a library, and this program is not really doing anything useful for the moment. I think it makes more sense to improve it a bit so that it becomes a sanity check on vendoring the library, as @dmcardle suggested below (if I didn't misunderstand that comment).

Eventually what I am hoping is to automate test generation by XKCP side and feed it to KMAC (from cryptolib, #16936). Since I do not know how CI works, I think I am not there yet (not sure if CI is the suitable place to test it either, in case it might break). Since KMAC supports 4 algorithms with varying-size inputs, having some randomly generated inputs at every run would make more sense for testing. We already have some [hard-coded tests](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/drivers/kmac_test_vectors.c) but their coverage is fairly limited.

File: third_party/xkcp/BUILD.xkcp.bazel
Comment (timothytrippel) [reply third_party/xkcp/BUILD.xkcp.bazel]: The fixes LGTM, thanks @ballifatih.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (timothytrippel) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: I see, so eventually you would potentially expand the tests here to use this XKCP library to generate random test vectors on the fly (presumably with some sort of seed so the test failures are reproducible)?

If that is the case, then perhaps there is no need for this host testing binary (`kmac_vector_generate`) att all? Could you can just have this PR vendor in the code (with Bazel as you have done) and then have a follow up PR use this library in a `opentitan_functest` that is an on-device test? You as long as the test does not have a  `long` or `eternal` timeout, it should run on the FPGA in CI.

File: third_party/xkcp/add_main_license.patch
Comment (timothytrippel) [inline third_party/xkcp/add_main_license.patch:9]: Hmm is this something you wrote? Do we need this at all?

Can't you just have a patch file that deletes all components that you do not use? Then you will only have vendored in what you need, and their respective LICENSE files.

Comment (ballifatih) [reply third_party/xkcp/add_main_license.patch:9]: That's also what I initially did but it has the following problem:

The main LICENSE file also has a GPL header for few files in the XKCP repo (we do not need these files so I excluded them in bazel config). When I make a patch on top of the original LICENSE file, the patch file then brings the GPL header into the patch file itself, since the diff file contains the lines removed from the original file. In short, since we do not bring any GPL-licensed file into the repo, I think it is best to avoid GPL headers as well (even if they are patches that *remove* those headers/files).

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: I removed the host binary `kmac_vector_generate` like you suggested. I agree that test generating can be better handled in a follow-up PR. This PR now only brings the XKCP bazel scripts.

Generating vectors directly on-device from a known seed could also work. Do you happen to know if it is feasible to generate long test vectors (e.g. something on the order of ~50 KB) directly on the device? I had just assumed that generating such long vectors on the device would not be feasible because of its runtime or memory budget. Although, come to think of it, generating such a large vector on the host and passing it to the device is also another challenge.

**Comment by ballifatih** [comment]
I received the green light regarding XKCP's licensing so I am marking this PR as ready for review/merge. 

The bazel scripts in this PR manage to bring only the part of XKCP that is relevant for our use-case, and these files are compatible with OT's license.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: Ah, I was confused because of the path. `sw/device` includes only device binaries along with tests (well some are host binaries because our unit tests run on the host). If this is a host helper/tool/utility binary it should go under `sw/host`. But I see that this file is no longer a part of this PR. Thanks @ballifatih, resolving this comment.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: [repating myself without repeating myself](https://github.com/lowRISC/opentitan/pull/16906#discussion_r1072785809)

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: I guess we should try and see. I'd prefer generating these on the device for the sake of simplicity.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: Resolving since this change is no longer a part of this PR.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: [.](https://github.com/lowRISC/opentitan/pull/16906#discussion_r1072790443).

File: third_party/xkcp/update_main_license.patch
Comment (alphan) [reply third_party/xkcp/update_main_license.patch]: [.](https://github.com/lowRISC/opentitan/pull/16906#issuecomment-1384027314)

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (alphan) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: [.](https://github.com/lowRISC/opentitan/pull/16906#discussion_r1072790443)

File: third_party/xkcp/add_main_license.patch
Comment (alphan) [reply third_party/xkcp/add_main_license.patch:9]: [.](https://github.com/lowRISC/opentitan/pull/16906#issuecomment-1384027314)

**Comment by alphan** [comment]
Thanks @ballifatih! I confirmed that all comments can now be resolved except for 2 forward-looking ones and the CI results are pretty recent.

File: sw/device/lib/testing/kmac/kmac_vector_generate.c
Comment (ballifatih) [reply sw/device/lib/testing/kmac/kmac_vector_generate.c]: Sounds good, I will try to see if we can generate KMAC vectors on the device in reasonable time. Thanks for suggestions @timothytrippel  and @alphan.

File: sw/device/lib/testing/kmac/BUILD
Comment (ballifatih) [reply sw/device/lib/testing/kmac/BUILD]: Since I removed the host binary from this PR, I added it to [TODO list](16410). I will come back to this later. Thanks @dmcardle.

**Pull request merged**
